src/rspq/rspq.c

```diff
@@ -0,0 +1,1235 @@
+    uint32_t state;             ///< Address of the overlay state in RDRAM (within data)
+    uint16_t code_size;         ///< Size of the code in bytes - 1
+    uint16_t data_size;         ///< Size of the data in bytes - 1
+} rspq_overlay_t;
+
+/// @cond
+_Static_assert(sizeof(rspq_overlay_t) == RSPQ_OVERLAY_DESC_SIZE);
+/// @endcond
+
+/**
+ * @brief The overlay table in DMEM. 
+ *
+ * This structure is defined in DMEM by rsp_queue.S, and contains the descriptors
+ * for the overlays, used by the queue engine to load each overlay when needed.
+ */
+typedef struct rspq_overlay_tables_s {
+    /** @brief Table mapping overlay ID to overlay index (used for the descriptors) */
+    uint8_t overlay_table[RSPQ_OVERLAY_TABLE_SIZE];
+    /** @brief Descriptor for each overlay, indexed by the previous table. */
+    rspq_overlay_t overlay_descriptors[RSPQ_MAX_OVERLAY_COUNT];
+} rspq_overlay_tables_t;
+
+/**
+ * @brief RSP Queue data in DMEM.
+ * 
+ * This structure is defined by rsp_queue.S, and represents the
+ * top portion of DMEM.
+ */
+typedef struct rsp_queue_s {
+    rspq_overlay_tables_t tables;        ///< Overlay table
+    /** @brief Pointer stack used by #RSPQ_CMD_CALL and #RSPQ_CMD_RET. */
+    uint32_t rspq_pointer_stack[RSPQ_MAX_BLOCK_NESTING_LEVEL];
+    uint32_t rspq_dram_lowpri_addr;      ///< Address of the lowpri queue (special slot in the pointer stack)
+    uint32_t rspq_dram_highpri_addr;     ///< Address of the highpri queue  (special slot in the pointer stack)
+    uint32_t rspq_dram_addr;             ///< Current RDRAM address being processed
+    int16_t current_ovl;                 ///< Current overlay index
+} __attribute__((aligned(16), packed)) rsp_queue_t;
+
+/**
+ * @brief RSP queue building context
+ * 
+ * This structure contains the state of a RSP queue as it is built by the CPU.
+ * It is instantiated two times: one for the lwopri queue, and one for the
+ * highpri queue. It contains the two buffers used in the double buffering
+ * scheme, and some metadata about the queue.
+ * 
+ * The current write pointer is stored in the "cur" field. The "sentinel" field
+ * contains the pointer to the last byte at which a new command can start,
+ * before overflowing the buffer (given #RSPQ_MAX_COMMAND_SIZE). This is used
+ * for efficiently check when it is time to switch to the other buffer: basically,
+ * it is sufficient to check whether "cur > sentinel".
+ * 
+ * The current queue is stored in 3 global pointers: #rspq_ctx, #rspq_cur_pointer
+ * and #rspq_cur_sentinel. #rspq_cur_pointer and #rspq_cur_sentinel are 
+ * external copies of the "cur" and "sentinel" pointer of the
+ * current context, but they are kept as separate global variables for
+ * maximum performance of the hottest code path: #rspq_write. In fact, it is
+ * much faster to access a global 32-bit pointer (via gp-relative offset) than
+ * dereferencing a member of a global structure pointer.
```

Comment:
How do we keep the ones in the struct in sync?

Reply:
The global pointers are saved back to the struct when switching context (see `rspq_switch_context`) and vice versa it updates the global pointers with the ones from the new context that is switched to. This happens when entering/exiting high-priority mode (see `rspq_highpri_begin` / `rspq_highpri_end`) or block mode (see `rspq_block_begin` / `rspq_block_end`).